src/app/services/country.service.ts

```diff
@@ -1,54 +1,56 @@
 import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { Observable, map, of, shareReplay } from 'rxjs';
+import { Observable, of, map, shareReplay, catchError } from 'rxjs';
 import { Country, CountryDetail } from '../models/country.model';
 
 @Injectable({
   providedIn: 'root',
 })
 export class CountryService {
   private apiUrl = 'https://restcountries.com/v3.1/all';
-  private countries$: Observable<Country[]>;
+  private searchByNameApiUrl = 'https://restcountries.com/v3.1/name'; // API for search by name
+  private searchByCapitalApiUrl = 'https://restcountries.com/v3.1/capital'; // API for search by capital
 
-  constructor(private http: HttpClient) {
-    this.countries$ = this.http
-      .get<Country[]>(this.apiUrl)
-      .pipe(shareReplay(1));
+  constructor(private http: HttpClient) {}
+  // Get all countries from cached API call
+  getAllCountries() {
+    return this.http.get<any>(this.apiUrl);
   }
```

Comment:
You should not call a method in the constructor of a service. It's considered a bad practice. Please remove this and call this inside the router effect or the resolver.